File: android/app/src/main/kotlin/com/privateagent/private_agent/model/TaskStep.kt
package com.privateagent.private_agent.model

import org.json.JSONObject
import java.util.regex.Pattern

data class TaskStep(
    val action: String,
    val params: Map<String, Any> = emptyMap(),
    val reasoning: String = ""
) {
    val isComplete: Boolean
        get() = action.equals(ACTION_DONE, ignoreCase = true)

    val signature: String
        get() = "$action:${params.entries.sortedBy { it.key }.joinToString(",") { "${it.key}=${it.value}" }}"

    fun getStringParam(key: String, default: String = ""): String {
        return params[key]?.toString() ?: default
    }

    fun getIntParam(key: String, default: Int = 0): Int {
        val value = params[key]
        return when (value) {
            is Number -> value.toInt()
            is String -> value.toDoubleOrNull()?.toInt() ?: default
            else -> default
        }
    }

    fun toJson(): JSONObject {
        val json = JSONObject()
        json.put("action", action)
        val paramsJson = JSONObject()
        params.forEach { (k, v) -> paramsJson.put(k, v) }
        json.put("params", paramsJson)
        if (reasoning.isNotBlank()) {
            json.put("reasoning", reasoning)
        }
        return json
    }

    companion object {
        const val ACTION_CLICK_TEXT = "click_text"
        const val ACTION_CLICK_AT = "click_at"
        const val ACTION_TYPE_TEXT = "type_text"
        const val ACTION_PRESS_ENTER = "press_enter"
        const val ACTION_SCROLL = "scroll"
        const val ACTION_SWIPE = "swipe"
        const val ACTION_PRESS_BACK = "press_back"
        const val ACTION_PRESS_HOME = "press_home"
        const val ACTION_OPEN_APP = "open_app"
        const val ACTION_WAIT = "wait"
        const val ACTION_DONE = "done"

        private val JSON_BLOCK_PATTERN = Pattern.compile("```(?:json)?\\s*([\\s\\S]*?)\\s*```", Pattern.CASE_INSENSITIVE)
        private val JSON_OBJECT_PATTERN = Pattern.compile("\\{[\\s\\S]*\\}")

        fun tryParse(raw: String): TaskStep? {
            val text = raw.trim()
            if (text.isEmpty()) return null

            // 1. Try markdown code fence
            val matcher = JSON_BLOCK_PATTERN.matcher(text)
            if (matcher.find()) {
                val block = matcher.group(1)?.trim().orEmpty()
                parseJsonString(block)?.let { return it }
            }

            // 2. Try raw text directly
            parseJsonString(text)?.let { return it }

            // 3. Try finding any top-level { ... } block
            val objMatcher = JSON_OBJECT_PATTERN.matcher(text)
            if (objMatcher.find()) {
                val candidate = objMatcher.group(0)?.trim().orEmpty()
                parseJsonString(candidate)?.let { return it }
            }

            return null
        }

        private fun parseJsonString(content: String): TaskStep? {
            return try {
                val json = JSONObject(content)
                val action = json.optString("action", "").trim()
                if (action.isEmpty()) return null

                val paramsMap = mutableMapOf<String, Any>()
                val paramsJson = json.optJSONObject("params")
                if (paramsJson != null) {
                    val keys = paramsJson.keys()
                    while (keys.hasNext()) {
                        val key = keys.next()
                        paramsMap[key] = paramsJson.get(key)
                    }
                }

                val reasoning = json.optString("reasoning", json.optString("thought", "")).trim()
                TaskStep(action = action, params = paramsMap, reasoning = reasoning)
            } catch (_: Exception) {
                null
            }
        }
    }
}

File: lib/models/task_step.dart
import 'dart:convert';

/// One atomic action decided by the LLM (or replayed from skill memory).
///
/// Wire format expected from the model:
/// ```json
/// {
///   "action": "click_text" | "click_at" | "type_text" | "press_enter" |
///             "scroll" | "swipe" | "press_back" | "press_home" |
///             "open_app" | "wait" | "done",
///   "params": { ... },
///   "reasoning": "short explanation",
///   "is_complete": false
/// }
/// ```
class TaskStep {
  static const actionClickText = 'click_text';
  static const actionClickAt = 'click_at';
  static const actionTypeText = 'type_text';
  static const actionPressEnter = 'press_enter';
  static const actionScroll = 'scroll';
  static const actionSwipe = 'swipe';
  static const actionPressBack = 'press_back';
  static const actionPressHome = 'press_home';
  static const actionOpenApp = 'open_app';
  static const actionWait = 'wait';
  static const actionDone = 'done';

  static const supportedActions = <String>{
    actionClickText,
    actionClickAt,
    actionTypeText,
    actionPressEnter,
    actionScroll,
    actionSwipe,
    actionPressBack,
    actionPressHome,
    actionOpenApp,
    actionWait,
    actionDone,
  };

  final String action;
  final Map<String, dynamic> params;
  final String reasoning;
  final bool isComplete;

  const TaskStep({
    required this.action,
    this.params = const {},
    this.reasoning = '',
    this.isComplete = false,
  });

  factory TaskStep.fromJson(Map<String, dynamic> json) {
    final rawAction = (json['action'] ?? '').toString().trim().toLowerCase();
    final rawParams = json['params'];
    return TaskStep(
      action: rawAction,
      params: rawParams is Map
          ? rawParams.map((k, v) => MapEntry(k.toString(), v))
          : const {},
      reasoning: (json['reasoning'] ?? '').toString(),
      isComplete: json['is_complete'] == true || json['isComplete'] == true,
    );
  }

  Map<String, dynamic> toJson() => {
        'action': action,
        'params': params,
        'reasoning': reasoning,
        'is_complete': isComplete,
      };

  bool get isSupported => supportedActions.contains(action);

  /// Stable signature used by the recovery engine to detect repeated actions.
  String get signature => '$action:${jsonEncode(params)}';

  /// Numeric param with forgiving parsing (int / double / numeric string).
  int? intParam(String key) {
    final v = params[key];
    if (v is int) return v;
    if (v is double) return v.round();
    if (v is String) return int.tryParse(v) ?? double.tryParse(v)?.round();
    return null;
  }

  String? stringParam(String key) {
    final v = params[key];
    if (v == null) return null;
    final s = v.toString();
    return s.isEmpty ? null : s;
  }

  /// Robustly extracts the first JSON object from raw LLM output, tolerating
  /// markdown fences, leading prose and trailing commentary.
  static TaskStep? tryParse(String aiText) {
    var text = aiText.trim();
    // Strip markdown code fences if present.
    final fence = RegExp(r'```(?:json)?\s*([\s\S]*?)```', caseSensitive: false);
    final fenceMatch = fence.firstMatch(text);
    if (fenceMatch != null) text = fenceMatch.group(1)!.trim();

    Map<String, dynamic>? decode(String candidate) {
      try {
        final decoded = jsonDecode(candidate);
        if (decoded is Map<String, dynamic> && decoded.containsKey('action')) {
          return decoded;
        }
      } catch (_) {}
      return null;
    }

    var parsed = decode(text);
    if (parsed == null) {
      // Fall back to the widest balanced { ... } region.
      final start = text.indexOf('{');
      final end = text.lastIndexOf('}');
      if (start >= 0 && end > start) {
        parsed = decode(text.substring(start, end + 1));
      }
    }
    if (parsed == null) return null;
    final step = TaskStep.fromJson(parsed);
    return step.isSupported ? step : null;
  }

  @override
  String toString() =>
      'TaskStep($action, params=$params, complete=$isComplete)';
}
